source/utils/RandomUtils.cc

```diff
@@ -0,0 +1,48 @@
+// ----------------------------------------------------------------------------
+// nexus | RandomUtils.cc
+//
+// This class defines commonly used functions to generate random quantities.
+//
+// The NEXT Collaboration
+// ----------------------------------------------------------------------------
+#include "RandomUtils.h"
+
+#include <Randomize.hh>
+
+#include "CLHEP/Units/SystemOfUnits.h"
+
+namespace nexus {
+
+  G4double RandomEnergy(G4double energy_max,G4double energy_min)
+  {
+    if (energy_max == energy_min){
+      return energy_min;
+    } else {
+      return (G4UniformRand()*(energy_max - energy_min) + energy_min);
+    }
+  }
+
+
+  G4ThreeVector Direction(G4double costheta_min, G4double costheta_max,
+                          G4double phi_min, G4double phi_max)
+  {
+    while (true) {
+      G4double cosTheta  = 2.*G4UniformRand()-1.;
+      if (cosTheta > costheta_min && cosTheta < costheta_max){
```

Comment:
This loop-until-we-hit approach to restricting the range of random number is a bit perverse, and it's a waste of CPU cycles.

Is there any good reason why we can't just map `G4UniformRandom`'s `[0,1]` range to whatever `[mix, max]` range we need? Line 30 already maps it to `[-1,1]` so why not just do the correct mapping in the first place? As it stands, we generate random numbers in a range that is too large and keep trying until we hit our smaller target. If the target is much smaller than the range, it will, on average, take lots and lots of attempts.

Is there *any* point to doing it this bizarre way?

Reply:
I think that what you propose works. It would be replacing line 30 with 

```
G4double cosTheta  = (costheta_max - costheta_min) * G4UniformRand() + costheta_min;
```

And the same with `phi`. This way we can simply remove the while loop whatsoever, is that right?

Reply:
Yes, that's exactly right. (Removing *both* `while` loops.)

Reply:
Hehe ... and when we're finished, of course, you should propagate (copy-paste) all we've learned here to the the other copy we have of this code on the `master` branch! (Sorry to bang on about it, but I must emphasize that the branch divergence is technical debt on which we're paying interest *all the time*.)